File: 207-course-schedule/207-course-schedule.py
from collections import defaultdict


def isCyclic(adj,visited,curr)->bool:
        
        if(visited[curr]==2): return True
    
        visited[curr]=2
        
        for i in range(len(adj[curr])):
            if(visited[adj[curr][i]]!=1):
                if(isCyclic(adj,visited,adj[curr][i])):
                    return True
                
            
        visited[curr]=1
        return False;
    

class Solution:
    
    
        
    
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        adj = defaultdict(list) # eg: of type adj = {0:[1],1:[0]}
        
        for i in prerequisites:
            adj[i[0]].append(i[1])
        
        visited = [0]*numCourses

        for i in range(0,numCourses):
            if(visited[i]==0):
                if(isCyclic(adj,visited,i)):
                    return False
            
        return True
               
        
        
        

File: 207-course-schedule/207-course-schedule.cpp
class Solution {
public:
    
    // 2 - > processing , 1-> processed, 0 ->not visited
    
    bool isCyclic( vector<vector<int>>& adj,vector<int>& visited,int curr){
        
        if(visited[curr]==2){
            return true;  // then cycle is there as u reach node which is processing 
        }
        
        visited[curr]=2; // not processing make it 
        
        for(int i=0;i<adj[curr].size();i++){
            if(visited[adj[curr][i]]!=1){  // skipping already processed nodes
                if(isCyclic(adj,visited,adj[curr][i]))
                    return true;
            } 
        }   
        
        visited[curr]=1;
        return false;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        
        vector<vector<int>> adj(numCourses);
        
        for(int i =0;i<prerequisites.size();i++){
            adj[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }
          
        vector<int> visited(numCourses,0);
        
        for(int i =0;i<numCourses;i++){
            if(visited[i]==0){
                if(isCyclic(adj,visited,i)){
                    return false;
                }
            }
        }
        
        return true;
        
        
    }
};